### ISSUE TITLE
Disble mouse when using keyboard as note input.

### ISSUE THREAD

**Issue by ErlingIv** [musescore/MuseScore#21399]
### Your idea

There should be an option to have the mouse working in "normal" mode when entering notes using the keyboard.
Now, when using the mouse to select a MuseScore option, I find myself "dragging" around with a note under the cursor.
Esc, solves that, but an option in preferences (note input?) would be of help.

### Problem to be solved

No problem, just an idea.

**Comment by Fastball2880** [comment]
Could you explain further?

Because there's already an option of this by using the "N" for note input, then change the pitch by using the arrow keys. You could also navigate the measure using "Ctrl+Left and Right." (Reference - https://musescore.org/en/handbook/4/default-keyboard-shortcuts#score-navigation)

**Comment by MarcSabatella** [comment]
The idea as discussed [on the forums](https://musescore.org/en/comment/1228012#comment-1228012) and elsewhere is to allow one to click to select a new note input cursor location while still in note input mode.  So that you can enter a few notes in one location, scroll somewhere else, click a measure, and start entering notes there, all without leaving note input mode.  Currently, clicking always enters a note, so you have to leave note input mode before clicking the new measure, or else you need to be careful not to just click to select the location but to actually select duration and click the note you want to enter.  If you forget and just click a measure without leaving note input mode and carefully entering the correct note, it enters some random note.  Undo will repair the damage, but the cursor remains in the original location.

So, *if* you are often making the mistake of simply clicking a measure to move the cursor while in note input mode (and we've all done it, I'm sure), it can make sense to wonder if there are alternatives.  I see a few possible ways to improve this:

1. A complete redesign of note input that eliminates the note input / normal mode distinction and hopefully solves this problem along with it
2. An optional setting in preferences that simply disables note input by mouse, period
3. A special modifier (like Alt) to allow click or drag to note enter a note (btw, Alt current causes click to enter a rest - I did not know that!)
4. A change to the default behavior so that a click anywhere *other* than in the current measure does not get interpreted as a note input command but rather as a simple cursor position change

Of these, 1 seems a pipe dream and likely to cause major disruptions but could be worth pursuing if someone on the design team has really good ideas.  2 is how I think most people envision this working, and I personally think it's a bad idea, as you have to discover the option somehow, and then it's all or nothing after that, but it's harmless enough I guess.  3 was suggested also but it's also hard to discover.  4 is something that may have been suggested at some point years ago, but I don't really remember so I'll take credit for it anyhow :-) and give it my vote.

I rarely use mouse input myself so it's hard to say how 4 would be perceived - it's an extra click to enter a note in the cases where you *do* really intend to enter a note in the new location and not merely select it.  But it could be interesting to try, and it would be a trivial change to implement.

**Comment by worldwideweary** [comment]
Might be far-fetched, but an alternative solution could be—since the user is "going back" to the keyboard anyways—to have an assigned shortcut that would take the currently "hovered" position of the mouse-pointer while in note-entry mode and navigate to/select that position, preparing for note entry via keyboard. Probably not what the request literally wants, but it seemed slightly worth mentioning.

**Comment by Lestrad** [comment]
I don't see why entering a note via the keyboard should automatically switch the mouse over to being a note input device. I agree that if one chooses to use the (computer) keyboard to enter notes, one should be able to continue to use the mouse normally - that is, to move the insertion point and select notes and rests. Having to constantly press Escape to be able to use the mouse normally almost makes the program unusable for me. 
The way I see it, it's not really a question of _disabling_ the mouse, but rather of being able to choose how to use the mouse.

**Comment by Valdessir** [comment]
I agree with Lestrad as well. I use keyboard to indicate what duration of note I want, and use the mouse to input the pitches. If I input the wrong note, I simply either press Ctrl+Z, shift+alt+up/down, or just up-down. In short, I use keyboard and mouse together, and would rather have the option to disable mouse on note input optional for users who prefer to only use keyboard all the way.

**Comment by MarcSabatella** [comment]
I think we all agree that there is room for improvement.  But please read over my previous comments to understand why the proposed solution is *not* good.  To restate in more detail:

- requiring an option to enable this behavior makes it non-discoverable to new users who are the ones most affected 9consider, most users are very accustomed to leaving note input mode - it hardly makes the program "unusable")
- the proposed option it's all-or-nothing, meaning that one cannot enter some notes with keyboard and others with the mouse as is very common (for instance, when building chords top down)

To be clear, entering notes by keyboard doesn't " automatically switch the mouse over to being a note input device".  It's *entering note input mode* that causes the keyboard commands as well as mouse input to be dedicated to that very task - entering notes.  Whether you touch the keyboard or not has no bearing whatsoever on how the mouse behaves.

So again, there is agreement that it would be good to improve this, but can we please consider the discussion of alternate solutions that don't suffer from the two major disadvantages I described?

**Comment by Valdessir** [comment]
I agree with you Marc. It needs to be discussed further to design a better way of going about this request.

**Comment by Lestrad** [comment]
Thanks Marc. 
I think you misunderstand my suggestion when you say that "the proposed solution is not good." You seem to think that it consists of "disabling input by mouse." Not at all. Your proposal No. 2 would be better expressed as "turning input by mouse on or off." You say it would be "all-or-nothing." In fact, Note Input mode is what's all-or-nothing. With the version I have (3.7.0), anyway, entering a note with the computer keyboard DOES automatically turn Note Entry mode on, and the mouse becomes a note input device. What you say ("Whether you touch the keyboard or not has no bearing whatsoever on how the mouse behaves") is just not true, at least for my version. 

When you think about it, the computer keyboard is ALWAYS a note input device. The CDEFGAB keys ALWAYS enter a note. All I'm asking is that entering a note from the keyboard not automatically switch the mouse over to being a note input device. Which it definitely does with my version (if there's a version where that doesn't happen, please let me know!).

Why not have a key  - "M"? "S"? (I know "M" is already a default shortcut) - that toggles mouse note input? The way "N" currently toggles Note Input?

**Comment by MarcSabatella** [comment]
Like I said, there should be more discussion about how exactly this should work.  Merely saying "there should be an option" is not a design - it's the *beginning* of a discussion.

Again, it's simply not true to say that the computer keys have any effect whatsoever on mouse behavior.  They don't, period.  What changes the mouse behavior is *entering note input mode*.  You can enter note input mode in several different ways.  Even without ever touching the keyboard, entering note input mode makes the mouse an input device - that  is the whole purpose of note input mode.  So clearing up this confusion about how MsueScore actually works is another important aspect of the discussion that needs to happen.

The forum is really the better place to hash out all these discussions about possible designs - GitHub is better for once a consensus is reached and the description can contain a simple and clear description of the design that people have agreed upon as best.  I still would argue that making users discover an option is not as good a design as the one I proposed, and I don't understand why no one has addressed that suggestion at all.

**Comment by Lestrad** [comment]
Thanks, Marc. Your version of MuseScore clearly doesn't act the same as mine. What version do you use? 

> "clearing up this **confusion** about how MsueScore actually works [...] Again, it's simply not true to say that the **computer keys** have any effect whatsoever on mouse behavior. They **don't**, period. 


Again, unless your version is different from mine (and _please_ tell me if it is and which version it is!), They do. For starters the "N" key toggles Note Input mode, and that changes the mouse behaviour radically. It becomes a note input device. And when I press  the A, B, C, D, E, F or G keys, Note Input Mode is entered automatically. I don't press "N" first or click the "N" icon or do any of the other things that turn on Note Input mode. **It turns on all by itself** when I enter a note from the computer keyboard. I am not at all confused about this. Again, if your version is different, please let me know.

> "no one has addressed that suggestion at all" 

No. 4? I did address it, in the forum. I said I'd be happy to have it:
> Having said that, your no. 4:
> "4 - A change to the default behavior so that a click anywhere other than in the current measure does not get interpreted as a note input command but rather as a simple cursor position change"
> would be fine. (https://musescore.org/en/comment/1228012#comment-1228012)

Rather than address other aspects of your last post, here I'll start a new thread on the Forum. I hope you'll give it a look.

Regards,
Les

**Comment by MarcSabatella** [comment]
I assure you that all versions of MuseScore behave the same.  Once again, the keyboard is *not* what triggers the change in mouse behavior - it is the entry to note input mode that does, whether initiated by keyboard or any other means.  In other words, the change is not connected to use of the keyboard *specifically*.

Feel free to include a link to your forum thread here.

**Comment by Lestrad** [comment]
Thanks. I'm in the process of writing up my contribution to a new thread. But please answer this: If what you say is true, why do I **not** have to press "N," or click the leftmost icon, or whatever, to enter Note Input mode before I enter a note from the keyboard? You say " whether initiated by keyboard or any other means." By "initiated by keyboard" do you mean pressing "N"? Or do you also mean pressing ABCDEF or G? Because I assure you - and I wish you'd give it a quick try to see if your version is the same - that all I have to do is press "A" and the mouse switches to note input.

**Comment by MarcSabatella** [comment]
Yes, in normal mode, pressing a key A-G has the side effect of enabling note input mode, as you can easily see by looking at the toolbar, and also by realizing that from that point forward, all other commands *also* start to function in the way they do when in note input mode.  So again, the point is, it doesn't matter *how* you enter note input mode - pressing N, clicking the note input icon, pressing a letter A-G, using a MIDI controller, etc - it's the *act of entering note input mode" that causes the mouse to start doing exactly what it is designed to do while in that mode - enter notes.

**Comment by Lestrad** [comment]
I've tried to explain why I have trouble with this in the [new thread](https://musescore.org/en/node/360642). I hope you'll give it a look. :-)

Regards,
Les

**Comment by Jojo-Schmitz** [comment]
Check my 3.7, which merged 2 commits from @worldwideweary: https://github.com/Jojo-Schmitz/MuseScore/pull/330 (it does fully contain https://github.com/Jojo-Schmitz/MuseScore/pull/323)
